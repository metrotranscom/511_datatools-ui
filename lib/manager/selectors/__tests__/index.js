// @flow

import {
  getActiveProject,
  getFilteredStops,
  getPatternData,
  getRouteData,
  getTimetableData
} from '../'

import mockData from '../../../../__tests__/test-utils/mock-data'

const {getMockInitialState} = mockData.store

// this mocked data models the data received from calling the
// gtfs.actions.timetables.fetchPatterns action
const mockPatternData = {
  patterns: [
    {
      pattern_id: '248',
      name:
        '10 stops from 1ST & METRO @ LR STATION to SANTA CLARA CALTRAIN STATION (133 trips)',
      shape: [
        {
          lat: 37.369805,
          lon: -121.9158
        },
        {
          lat: 37.369491,
          lon: -121.915578
        }
      ],
      stops: [
        {
          stop_id: '20'
        },
        {
          stop_id: '5228'
        }
      ],
      trips: [
        {
          frequencies: [],
          stop_times: [
            {
              arrival_time: 23580,
              departure_time: 23580
            }
          ]
        },
        {
          frequencies: [],
          stop_times: [
            {
              arrival_time: 76500,
              departure_time: 76500
            }
          ]
        }
      ],
      geometry: {
        type: 'LineString',
        coordinates: [[-121.9158, 37.369805], [-121.915578, 37.369491]]
      },
      route_id: '10',
      route_name: '10 - SANTA CLARA TRANSIT- METRO AIRPORT'
    }
  ],
  stops: [
    {
      location_type: null,
      stop_code: '60020',
      stop_desc: null,
      stop_id: '20',
      stop_lat: 37.355056803,
      stop_lon: -121.945134661,
      stop_name: 'EL CAMINO & LAFAYETTE',
      stop_url: null,
      wheelchair_boarding: null,
      zone_id: '1'
    },
    {
      location_type: null,
      stop_code: '65228',
      stop_desc: null,
      stop_id: '5228',
      stop_lat: 37.35339598,
      stop_lon: -121.937117102,
      stop_name: 'SANTA CLARA CALTRAIN STATION',
      stop_url: null,
      wheelchair_boarding: null,
      zone_id: '1'
    }
  ]
}

describe('manager > selectors >', () => {
  describe('getActiveProject', () => {
    it('should get the project', () => {
      const mockProject = {
        id: '1',
        autoFetchFeeds: false,
        autoFetchHour: 0,
        autoFetchMinute: 0,
        bounds: null,
        defaultTimeZone: null,
        lastUpdated: 1546955864840,
        name: 'Test Project',
        organizationId: null,
        otpServers: null,
        useCustomOsmBounds: false
      }
      const mockState = getMockInitialState()
      // $FlowFixMe - ignore type misalignment for testing purposes
      mockState.projects.active = mockProject
      // $FlowFixMe - ignore type misalignment for testing purposes
      mockState.projects.all = [mockProject]
      expect(getActiveProject(mockState)).toMatchSnapshot()
    })
  })

  describe('getFilteredStops', () => {
    it('should output blank array when no pattern data exists', () => {
      expect(getFilteredStops(getMockInitialState())).toMatchSnapshot()
    })

    it('should output data when pattern data exists', () => {
      const mockState = getMockInitialState()
      mockState.gtfs.patterns.data = mockPatternData

      expect(getFilteredStops(mockState)).toMatchSnapshot()
    })
  })

  // WARNING: these are unit tests where the output of the function being tested
  // is very dependent on downstream implementation
  describe('getRouteDetailsData', () => {
    it('should output blank array when no route data exists', () => {
      expect(getRouteData(getMockInitialState())).toMatchSnapshot()
    })

    it('should output data when route data exists', () => {
      const mockState = getMockInitialState()
      mockState.gtfs.routes.routeDetails.data = {
        numRoutes: 1,
        routes: [
          {
            route_id: '10',
            route_short_name: '10',
            route_long_name: 'SANTA CLARA TRANSIT- METRO AIRPORT',
            route_desc: null,
            stops: [
              {
                stop_id: '20'
              },
              {
                stop_id: '5228'
              }
            ],
            trips: [
              {
                frequencies: [],
                pattern_id: '284',
                stop_times: [
                  {
                    arrival_time: 23820,
                    departure_time: 23820
                  }
                ]
              }
            ]
          }
        ]
      }

      expect(getRouteData(mockState)).toMatchSnapshot()
    })
  })

  // WARNING: these are unit tests where the output of the function being tested
  // is very dependent on downstream implementation
  describe('getPatternData', () => {
    it('should output blank array when no pattern data exists', () => {
      expect(getPatternData(getMockInitialState())).toMatchSnapshot()
    })

    it('should output data when pattern data exists', () => {
      const mockState = getMockInitialState()
      mockState.gtfs.patterns.data = mockPatternData

      expect(getPatternData(mockState)).toMatchSnapshot()
    })
  })

  // WARNING: these are unit tests where the output of the function being tested
  // is very dependent on downstream implementation
  describe('getTimetableData', () => {
    // this mocked data models the data received from calling the
    // gtfs.actions.timetables.fetchTimetablesWithFilters action
    const mockTimetableData = {
      feed: {
        patterns: [
          {
            stops: [
              {stop_id: '15', stop_name: 'TECHNOLOGY & METRO'},
              {stop_id: '14', stop_name: '1ST & METRO @ LR STATION'}
            ],
            trips: [
              {
                direction_id: 1,
                pattern_id: '248',
                service_id: 'Weekdays',
                stop_times: [
                  {
                    arrival_time: 23580,
                    departure_time: 23580,
                    stop_id: '14',
                    stop_sequence: 1,
                    timepoint: null
                  },
                  {
                    arrival_time: null,
                    departure_time: null,
                    stop_id: '15',
                    stop_sequence: 2,
                    timepoint: null
                  }
                ],
                trip_headsign: '10 SANTA CLARA CALTRAIN VIA AIRPORT',
                trip_id: '2369269',
                trip_short_name: null
              },
              {
                direction_id: 1,
                pattern_id: '248',
                service_id: 'Weekdays',
                stop_times: [
                  {
                    arrival_time: 28980,
                    departure_time: 28980,
                    stop_id: '14',
                    stop_sequence: 1,
                    timepoint: null
                  },
                  {
                    arrival_time: null,
                    departure_time: null,
                    stop_id: '15',
                    stop_sequence: 2,
                    timepoint: null
                  }
                ],
                trip_headsign: '10 SANTA CLARA CALTRAIN VIA AIRPORT',
                trip_id: '2369292',
                trip_short_name: null
              }
            ]
          }
        ]
      }
    }

    it('should get the data when no data is fetched', () => {
      expect(getTimetableData(getMockInitialState())).toMatchSnapshot()
    })

    it('should get the data when data is available', () => {
      const mockState = getMockInitialState()
      mockState.gtfs.timetables.data = mockTimetableData
      expect(getTimetableData(mockState)).toMatchSnapshot()
    })

    it('should get the data when data is available and timepoint filter is set', () => {
      const mockState = getMockInitialState()
      mockState.gtfs.filter.timepointFilter = true
      mockState.gtfs.timetables.data = mockTimetableData
      expect(getTimetableData(mockState)).toMatchSnapshot()
    })

    it('should get the data when data is available and show arrivals filter is set', () => {
      const mockState = getMockInitialState()
      mockState.gtfs.filter.showArrivals = true
      mockState.gtfs.timetables.data = mockTimetableData
      expect(getTimetableData(mockState)).toMatchSnapshot()
    })
  })
})
