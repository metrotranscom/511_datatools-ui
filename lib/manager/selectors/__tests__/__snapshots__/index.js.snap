// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`manager > selectors > getActiveProject should get the project 1`] = `
Object {
  "autoFetchFeeds": false,
  "autoFetchHour": 0,
  "autoFetchMinute": 0,
  "bounds": null,
  "defaultTimeZone": null,
  "id": "1",
  "lastUpdated": 1546955864840,
  "name": "Test Project",
  "organizationId": null,
  "otpServers": null,
  "useCustomOsmBounds": false,
}
`;

exports[`manager > selectors > getFilteredStops should output blank array when no pattern data exists 1`] = `Array []`;

exports[`manager > selectors > getFilteredStops should output data when pattern data exists 1`] = `
Array [
  Object {
    "location_type": null,
    "stop_code": "60020",
    "stop_desc": null,
    "stop_id": "20",
    "stop_lat": 37.355056803,
    "stop_lon": -121.945134661,
    "stop_name": "EL CAMINO & LAFAYETTE",
    "stop_url": null,
    "wheelchair_boarding": null,
    "zone_id": "1",
  },
  Object {
    "location_type": null,
    "stop_code": "65228",
    "stop_desc": null,
    "stop_id": "5228",
    "stop_lat": 37.35339598,
    "stop_lon": -121.937117102,
    "stop_name": "SANTA CLARA CALTRAIN STATION",
    "stop_url": null,
    "wheelchair_boarding": null,
    "zone_id": "1",
  },
]
`;

exports[`manager > selectors > getPatternData should output blank array when no pattern data exists 1`] = `Array []`;

exports[`manager > selectors > getPatternData should output data when pattern data exists 1`] = `
Array [
  Object {
    "numStops": 2,
    "numTrips": 2,
    "patternId": "248",
    "patternName": "10 stops from 1ST & METRO @ LR STATION to SANTA CLARA CALTRAIN STATION (133 trips)",
    "tripsPerHour": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  },
]
`;

exports[`manager > selectors > getRouteDetailsData should output blank array when no route data exists 1`] = `Array []`;

exports[`manager > selectors > getRouteDetailsData should output data when route data exists 1`] = `
Array [
  Object {
    "numPatterns": 1,
    "numStops": 2,
    "numTrips": 1,
    "routeId": "10",
    "routeName": "10 - SANTA CLARA TRANSIT- METRO AIRPORT",
    "tripsPerHour": Array [
      0,
      0,
      0,
      0,
      0,
      0,
      1,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
      0,
    ],
  },
]
`;

exports[`manager > selectors > getTimetableData should get the data when data is available 1`] = `
Object {
  "columns": Array [
    Object {
      "children": "Trip ID",
      "dataField": "tripId",
      "hidden": true,
      "isKey": true,
      "row": "0",
      "rowSpan": "2",
      "width": "150",
    },
    Object {
      "children": "Trip Headsign",
      "dataField": "tripHeadsign",
      "hidden": true,
      "row": "0",
      "rowSpan": "2",
      "width": "150",
    },
    Object {
      "children": "Trip Short Name",
      "dataField": "tripShortName",
      "hidden": true,
      "row": "0",
      "rowSpan": "2",
      "width": "150",
    },
    Object {
      "children": <span>
        1ST & METRO @ LR STATION
      </span>,
      "colSpan": "2",
      "csvHeader": "1ST & METRO @ LR STATION",
      "dataField": undefined,
      "dataFormat": [Function],
      "formatExtraData": Array [
        Array [
          Object {
            "count": 1,
            "diff": NaN,
          },
          Object {
            "count": 1,
            "diff": 5400,
          },
        ],
        Array [
          Object {
            "count": 2,
            "diff": null,
          },
        ],
      ],
      "headerText": "1ST & METRO @ LR STATION",
      "hidden": undefined,
      "row": "0",
      "stopId": "14",
      "width": "170",
    },
    Object {
      "children": "Arrive",
      "csvHeader": "Arrive",
      "dataField": "0-arrival",
      "dataFormat": [Function],
      "formatExtraData": Array [
        Array [
          Object {
            "count": 1,
            "diff": NaN,
          },
          Object {
            "count": 1,
            "diff": 5400,
          },
        ],
        Array [
          Object {
            "count": 2,
            "diff": null,
          },
        ],
      ],
      "hidden": undefined,
      "row": "1",
      "stopId": "14",
      "width": "85",
    },
    Object {
      "children": "Dwell",
      "csvHeader": "Dwell",
      "dataField": "0-dwell",
      "hidden": undefined,
      "row": "1",
      "stopId": "14",
      "width": "85",
    },
    Object {
      "children": <span>
        TECHNOLOGY 
        <br />
         METRO
      </span>,
      "colSpan": "2",
      "csvHeader": "TECHNOLOGY & METRO",
      "dataField": undefined,
      "dataFormat": [Function],
      "formatExtraData": Array [
        Array [
          Object {
            "count": 1,
            "diff": NaN,
          },
          Object {
            "count": 1,
            "diff": 5400,
          },
        ],
        Array [
          Object {
            "count": 2,
            "diff": null,
          },
        ],
      ],
      "headerText": "TECHNOLOGY & METRO",
      "hidden": undefined,
      "row": "0",
      "stopId": "15",
      "width": "170",
    },
    Object {
      "children": "Arrive",
      "csvHeader": "Arrive",
      "dataField": "1-arrival",
      "dataFormat": [Function],
      "formatExtraData": Array [
        Array [
          Object {
            "count": 1,
            "diff": NaN,
          },
          Object {
            "count": 1,
            "diff": 5400,
          },
        ],
        Array [
          Object {
            "count": 2,
            "diff": null,
          },
        ],
      ],
      "hidden": undefined,
      "row": "1",
      "stopId": "15",
      "width": "85",
    },
    Object {
      "children": "Dwell",
      "csvHeader": "Dwell",
      "dataField": "1-dwell",
      "hidden": undefined,
      "row": "1",
      "stopId": "15",
      "width": "85",
    },
  ],
  "rows": Array [
    Object {
      "0-arrival": 23580,
      "0-departure": 23580,
      "0-dwell": "--",
      "1-arrival": null,
      "1-departure": null,
      "1-dwell": "--",
      "tripHeadsign": "10 SANTA CLARA CALTRAIN VIA AIRPORT",
      "tripId": "2369269",
      "tripShortName": null,
    },
    Object {
      "0-arrival": 28980,
      "0-departure": 28980,
      "0-dwell": "--",
      "1-arrival": null,
      "1-departure": null,
      "1-dwell": "--",
      "tripHeadsign": "10 SANTA CLARA CALTRAIN VIA AIRPORT",
      "tripId": "2369292",
      "tripShortName": null,
    },
  ],
}
`;

exports[`manager > selectors > getTimetableData should get the data when data is available and show arrivals filter is set 1`] = `
Object {
  "columns": Array [
    Object {
      "children": "Trip ID",
      "dataField": "tripId",
      "hidden": true,
      "isKey": true,
      "row": "0",
      "rowSpan": "2",
      "width": "150",
    },
    Object {
      "children": "Trip Headsign",
      "dataField": "tripHeadsign",
      "hidden": true,
      "row": "0",
      "rowSpan": "2",
      "width": "150",
    },
    Object {
      "children": "Trip Short Name",
      "dataField": "tripShortName",
      "hidden": true,
      "row": "0",
      "rowSpan": "2",
      "width": "150",
    },
    Object {
      "children": <span>
        1ST & METRO @ LR STATION
      </span>,
      "colSpan": "2",
      "csvHeader": "1ST & METRO @ LR STATION",
      "dataField": undefined,
      "dataFormat": [Function],
      "formatExtraData": Array [
        Array [
          Object {
            "count": 1,
            "diff": NaN,
          },
          Object {
            "count": 1,
            "diff": 5400,
          },
        ],
        Array [
          Object {
            "count": 2,
            "diff": null,
          },
        ],
      ],
      "headerText": "1ST & METRO @ LR STATION",
      "hidden": undefined,
      "row": "0",
      "stopId": "14",
      "width": "170",
    },
    Object {
      "children": "Arrive",
      "csvHeader": "Arrive",
      "dataField": "0-arrival",
      "dataFormat": [Function],
      "formatExtraData": Array [
        Array [
          Object {
            "count": 1,
            "diff": NaN,
          },
          Object {
            "count": 1,
            "diff": 5400,
          },
        ],
        Array [
          Object {
            "count": 2,
            "diff": null,
          },
        ],
      ],
      "hidden": undefined,
      "row": "1",
      "stopId": "14",
      "width": "85",
    },
    Object {
      "children": "Dwell",
      "csvHeader": "Dwell",
      "dataField": "0-dwell",
      "hidden": undefined,
      "row": "1",
      "stopId": "14",
      "width": "85",
    },
    Object {
      "children": <span>
        TECHNOLOGY 
        <br />
         METRO
      </span>,
      "colSpan": "2",
      "csvHeader": "TECHNOLOGY & METRO",
      "dataField": undefined,
      "dataFormat": [Function],
      "formatExtraData": Array [
        Array [
          Object {
            "count": 1,
            "diff": NaN,
          },
          Object {
            "count": 1,
            "diff": 5400,
          },
        ],
        Array [
          Object {
            "count": 2,
            "diff": null,
          },
        ],
      ],
      "headerText": "TECHNOLOGY & METRO",
      "hidden": undefined,
      "row": "0",
      "stopId": "15",
      "width": "170",
    },
    Object {
      "children": "Arrive",
      "csvHeader": "Arrive",
      "dataField": "1-arrival",
      "dataFormat": [Function],
      "formatExtraData": Array [
        Array [
          Object {
            "count": 1,
            "diff": NaN,
          },
          Object {
            "count": 1,
            "diff": 5400,
          },
        ],
        Array [
          Object {
            "count": 2,
            "diff": null,
          },
        ],
      ],
      "hidden": undefined,
      "row": "1",
      "stopId": "15",
      "width": "85",
    },
    Object {
      "children": "Dwell",
      "csvHeader": "Dwell",
      "dataField": "1-dwell",
      "hidden": undefined,
      "row": "1",
      "stopId": "15",
      "width": "85",
    },
  ],
  "rows": Array [
    Object {
      "0-arrival": 23580,
      "0-departure": 23580,
      "0-dwell": "--",
      "1-arrival": null,
      "1-departure": null,
      "1-dwell": "--",
      "tripHeadsign": "10 SANTA CLARA CALTRAIN VIA AIRPORT",
      "tripId": "2369269",
      "tripShortName": null,
    },
    Object {
      "0-arrival": 28980,
      "0-departure": 28980,
      "0-dwell": "--",
      "1-arrival": null,
      "1-departure": null,
      "1-dwell": "--",
      "tripHeadsign": "10 SANTA CLARA CALTRAIN VIA AIRPORT",
      "tripId": "2369292",
      "tripShortName": null,
    },
  ],
}
`;

exports[`manager > selectors > getTimetableData should get the data when data is available and timepoint filter is set 1`] = `
Object {
  "columns": Array [
    Object {
      "children": "Trip ID",
      "dataField": "tripId",
      "hidden": true,
      "isKey": true,
      "row": "0",
      "rowSpan": "2",
      "width": "150",
    },
    Object {
      "children": "Trip Headsign",
      "dataField": "tripHeadsign",
      "hidden": true,
      "row": "0",
      "rowSpan": "2",
      "width": "150",
    },
    Object {
      "children": "Trip Short Name",
      "dataField": "tripShortName",
      "hidden": true,
      "row": "0",
      "rowSpan": "2",
      "width": "150",
    },
    Object {
      "children": <span>
        1ST & METRO @ LR STATION
      </span>,
      "colSpan": "2",
      "csvHeader": "1ST & METRO @ LR STATION",
      "dataField": undefined,
      "dataFormat": [Function],
      "formatExtraData": Array [
        Array [
          Object {
            "count": 1,
            "diff": NaN,
          },
          Object {
            "count": 1,
            "diff": 5400,
          },
        ],
        Array [
          Object {
            "count": 2,
            "diff": null,
          },
        ],
      ],
      "headerText": "1ST & METRO @ LR STATION",
      "hidden": false,
      "row": "0",
      "stopId": "14",
      "width": "170",
    },
    Object {
      "children": "Arrive",
      "csvHeader": "Arrive",
      "dataField": "0-arrival",
      "dataFormat": [Function],
      "formatExtraData": Array [
        Array [
          Object {
            "count": 1,
            "diff": NaN,
          },
          Object {
            "count": 1,
            "diff": 5400,
          },
        ],
        Array [
          Object {
            "count": 2,
            "diff": null,
          },
        ],
      ],
      "hidden": false,
      "row": "1",
      "stopId": "14",
      "width": "85",
    },
    Object {
      "children": "Dwell",
      "csvHeader": "Dwell",
      "dataField": "0-dwell",
      "hidden": false,
      "row": "1",
      "stopId": "14",
      "width": "85",
    },
    Object {
      "children": <span>
        TECHNOLOGY 
        <br />
         METRO
      </span>,
      "colSpan": "2",
      "csvHeader": "TECHNOLOGY & METRO",
      "dataField": undefined,
      "dataFormat": [Function],
      "formatExtraData": Array [
        Array [
          Object {
            "count": 1,
            "diff": NaN,
          },
          Object {
            "count": 1,
            "diff": 5400,
          },
        ],
        Array [
          Object {
            "count": 2,
            "diff": null,
          },
        ],
      ],
      "headerText": "TECHNOLOGY & METRO",
      "hidden": true,
      "row": "0",
      "stopId": "15",
      "width": "170",
    },
    Object {
      "children": "Arrive",
      "csvHeader": "Arrive",
      "dataField": "1-arrival",
      "dataFormat": [Function],
      "formatExtraData": Array [
        Array [
          Object {
            "count": 1,
            "diff": NaN,
          },
          Object {
            "count": 1,
            "diff": 5400,
          },
        ],
        Array [
          Object {
            "count": 2,
            "diff": null,
          },
        ],
      ],
      "hidden": true,
      "row": "1",
      "stopId": "15",
      "width": "85",
    },
    Object {
      "children": "Dwell",
      "csvHeader": "Dwell",
      "dataField": "1-dwell",
      "hidden": true,
      "row": "1",
      "stopId": "15",
      "width": "85",
    },
  ],
  "rows": Array [
    Object {
      "0-arrival": 23580,
      "0-departure": 23580,
      "0-dwell": "--",
      "1-arrival": null,
      "1-departure": null,
      "1-dwell": "--",
      "tripHeadsign": "10 SANTA CLARA CALTRAIN VIA AIRPORT",
      "tripId": "2369269",
      "tripShortName": null,
    },
    Object {
      "0-arrival": 28980,
      "0-departure": 28980,
      "0-dwell": "--",
      "1-arrival": null,
      "1-departure": null,
      "1-dwell": "--",
      "tripHeadsign": "10 SANTA CLARA CALTRAIN VIA AIRPORT",
      "tripId": "2369292",
      "tripShortName": null,
    },
  ],
}
`;

exports[`manager > selectors > getTimetableData should get the data when no data is fetched 1`] = `
Object {
  "columns": Array [],
  "rows": Array [],
}
`;
